### PR TITLE
Add Toggle Param to Return Implied Scopes from Remix API Query Method

### PR DESCRIPTION
**Pull request by allenchazhoor**
### Issue

Closes https://github.com/Shopify/develop-app-runtime-primitives/issues/507

### Context

Currently [the query method in the Remix API](https://github.com/Shopify/shopify-app-js/blob/eefac27d6857ee2f33fb8a5b4925edc9545f61bc/packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/query.ts#L44) does not return implied scopes but the [response from Core](https://github.com/Shopify/shopify-app-js/blob/eefac27d6857ee2f33fb8a5b4925edc9545f61bc/packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/query.ts#L32) does. This is because Remix is returning a [compressed set of scopes](https://github.com/Shopify/shopify-app-js/blob/eefac27d6857ee2f33fb8a5b4925edc9545f61bc/packages/apps/shopify-api/lib/auth/scopes/index.ts#L83) from the response from Core, using [the toArray() method in AuthScopes](https://github.com/Shopify/shopify-app-js/blob/eefac27d6857ee2f33fb8a5b4925edc9545f61bc/packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/query.ts#L45).

This PR adds a parameter in the AuthScopes toArray() method to toggle returning the original set of scopes returned from from the Admin API.

Tophatted here:
https://admin.web.partners-wppf.allen-chazhoor.us.spin.dev/store/development-store-2/apps/querytophat
Use the DevUI or the app itself to revoke/request scopes
image
Here we can see that when using the remix query, we now return the implied scope `read_customers`

## Type of change

- [x] Patch: Bug (non-breaking change which fixes an issue)
- [ ] Minor: New feature (non-breaking change which adds functionality)
- [ ] Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have used `pnpm changeset` to create a draft changelog entry (do NOT update the `CHANGELOG.md` files manually)
- [x] I have added/updated tests for this change
- [ ] I have documented new APIs/updated the documentation for modified APIs (for public APIs)

### CODE REVIEW

**Review by rezaansyed — APPROVED** [review]
LGTM!

File: packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/query.test.ts [hunk 1]
```diff
@@ -25,9 +25,13 @@ it('returns scopes information', async () => {
...
   expect(result).not.toBeUndefined();
-  expect(result.granted).toEqual(['read_orders', 'write_customers']);
+  expect(result.granted).toEqual([
+    'read_orders',
+    'write_customers',
+    'read_customers',
+  ]);
   expect(result.required).toEqual(['read_orders', 'read_reports']);
-  expect(result.optional).toEqual(['write_customers']);
+  expect(result.optional).toEqual(['write_customers', 'read_customers']);
```
[the remark this answers is not in the document]
Comment (RyanDJLee) [reply packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/query.test.ts:34]: It may be helpful to revisit our [mocked responses](https://github.com/Shopify/shopify-app-js/blob/7635b71da2c18ea2fc15c7b70ed94512207e9a2b/packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/mock-responses.ts) as well.

[the remark this answers is not in the document]
Comment (allenchazhoor) [reply packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/query.test.ts:34]: While updating coverage, I realized there might be an edge case. Do we wish to return required scopes as implied optional ones as well? 

For example, if we request explicitly for "read_customers" and set "write customers" as an optional scope, and then both are granted. Under the changes I have made, we would return "read_customers" in `required` but also as an implicit scope in `optional`.

File: packages/apps/shopify-api/lib/auth/scopes/index.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ class AuthScopes {
 
   private compressedScopes: Set<string>;
   private expandedScopes: Set<string>;
+  private originalScopes: Set<string>;
```
Comment (RyanDJLee) [inline packages/apps/shopify-api/lib/auth/scopes/index.ts:9]: Should we consider how the other methods are implemented as well?

The current behaviour considers AuthScopes(['read_products']) equivalent to AuthScopes(['read_products', 'write_products']). Since we are encouraging developers to use this class, I wonder if this behaviour is consistent with the changes proposed in this PR / as part of the Optional scopes feature?

File: packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/query.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (allenchazhoor) [reply packages/apps/shopify-app-remix/src/server/authenticate/admin/scope/__tests__/query.test.ts]: I have corrected the Remix query behaviour to return exactly what Admin API returns. 
 
 I have also corrected the original test coverage and mock calls and added new scopes to demonstrate functionality.
 
 - `read_reports` has been added to the `installedAccessScopes` as it is a required scope
 -  `read_customers` has been added to the `installedAccessScopes` as it is a implied scope of the installed `write_customers`
 -   `write_products` has been added to `optionalAccessScopes` and `read_products` has been added to `requiredAccessScopes` and `installedAccessScopes`, to ensure that, like with the Admin API, the Remix API does not return required scopes as implied in optional scopes.
 
image

 
image

File: packages/apps/shopify-api/lib/auth/scopes/index.ts (hunk 2 above)
Comment (RyanDJLee) [reply packages/apps/shopify-api/lib/auth/scopes/index.ts:9]: Ideally, we would want to keep the scope of this change tight and not introduce breaking changes (by changing the default behaviour). However, I wonder about the developer use cases. Would they want to compare and check inclusion with respect to the exact set of scopes or the [expanded](https://github.com/Shopify/shopify-app-js/blob/9733952bfa79f65bde80df7b3c718ac2f67047c4/packages/apps/shopify-api/lib/auth/scopes/index.ts#L52) set of scopes? What about the other methods?


i.e.
```
required: ['read_products']
optional: ['write_products']
granted: ['read_products']
```

1. Developer queries for optional scopes and instantiates:
```
requiredScopes = AuthScope(['read_products']);
optionalScopes = AuthScope(['write_products']);
grantedScopes = AuthScope(['read_products']);
```
2. Developer wants to check whether the optional scopes have been granted:
```
grantedScopes.has(optionalScopes);
grantedScopes.equals(optionalScopes);
```

* Is this a realistic use case?
* **Is current implementation behaving as expected in these scenarios?**


cc @kbav

**Review by lizkenyon — APPROVED** [review]
Let's add another test, and add some details to the changelog. 

I think it would make sense to address the equality in a separate PR.

**Review by RyanDJLee — APPROVED** [review]

**Review by kbav — APPROVED** [review]
LGTM from what I'm seeing. Let's address CI issues and get this shipped!

Regarding @lizkenyon’s [semantic versioning comment](https://github.com/Shopify/shopify-app-js/pull/1604#discussion_r1803474090), some context from my perspective, this does seem to be a patch.
- This is indeed a bug / unintended response value. We want for this to return expanded / implied scopes, just like our GraphQL responses.
- Nothing will change for existing implementations or apps using `.toArray()` from `AuthScopes`
- While the scopes API’s `query` method's response _will_ change, this API's use is guarded behind the `wip_optionalScopesApi` future flag. My thought is that the "`wip_`" prefix is indicating to developers that this is not only "unstable", but potentially in active, breaking development (functionally untethered to SemVer major & minor until it’s no longer "wip")

If that sounds right to Liz & other folks, I'm on board shipping this as a patch!
